# Імпорти

In [2]:
from keras.datasets import imdb
from keras.utils import pad_sequences
from keras.models import Sequential
from keras.layers import Embedding, LSTM, Dense, Bidirectional, Dropout

# Задаємо гіперпараметри моделі

In [3]:
max_features = 10000  # Кількість слів, що використовуються для векторизації
maxlen = 200  # Максимальна довжина рецензії (обрізаємо або доповнюємо до цієї довжини)

# Завантаження даних

In [4]:
(x_train, y_train), (x_test, y_test) = imdb.load_data(num_words=max_features)

17464789/17464789 [==============================] - 0s 0us/step


# Попередня обробка даних

In [5]:
x_train = pad_sequences(x_train, maxlen=maxlen)
x_test = pad_sequences(x_test, maxlen=maxlen)

# Побудова моделі

In [6]:
model = Sequential()
model.add(Embedding(max_features, 128, input_length=maxlen))
model.add(Bidirectional(LSTM(64))) # Двосторонній LSTM шар з 64 нейронами
model.add(Dropout(0.5))
model.add(Dense(1, activation='sigmoid'))

# Компіляція моделі

In [7]:
model.compile(
    loss='binary_crossentropy',
    optimizer='rmsprop',
    metrics=['accuracy']
    )

# Виведення опису моделі

In [8]:
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding (Embedding)       (None, 200, 128)          1280000   
                                                                 
 bidirectional (Bidirectiona  (None, 128)              98816     
 l)                                                              
                                                                 
 dropout (Dropout)           (None, 128)               0         
                                                                 
 dense (Dense)               (None, 1)                 129       
                                                                 
Total params: 1,378,945
Trainable params: 1,378,945
Non-trainable params: 0
_________________________________________________________________


# Навчання моделі

In [9]:
batch_size = 32
epochs = 5

model.fit(
    x_train,
    y_train,
    batch_size=batch_size,
    epochs=epochs,
    validation_data=(x_test, y_test)
    )

Epoch 1/5
782/782 [==============================] - 341s 431ms/step - loss: 0.4480 - accuracy: 0.7876 - val_loss: 0.3560 - val_accuracy: 0.8468
Epoch 2/5
782/782 [==============================] - 333s 426ms/step - loss: 0.3009 - accuracy: 0.8783 - val_loss: 0.3101 - val_accuracy: 0.8713
Epoch 3/5
782/782 [==============================] - 333s 426ms/step - loss: 0.2551 - accuracy: 0.9016 - val_loss: 0.3889 - val_accuracy: 0.8282
Epoch 4/5
782/782 [==============================] - 335s 429ms/step - loss: 0.2229 - accuracy: 0.9178 - val_loss: 0.5091 - val_accuracy: 0.8304
Epoch 5/5
782/782 [==============================] - 334s 428ms/step - loss: 0.1989 - accuracy: 0.9268 - val_loss: 0.3238 - val_accuracy: 0.8784


# Оцінка результатів тренування

In [10]:
loss, accuracy = model.evaluate(x_test, y_test)
print(f'Loss: {loss:.4f}')
print(f'Accuracy: {accuracy*100:.2f}%')

782/782 [==============================] - 62s 79ms/step - loss: 0.3238 - accuracy: 0.8784
Loss: 0.3238
Accuracy: 87.84%


# Висновки

Модель з SimpleRNN:

Втрати після 5 епох: 0.3991
Точність після 5 епох: 0.8290
Втрати на валідації після 5 епох: 0.6005
Точність на валідації після 5 епох: 0.7178
Модель з LSTM:

Втрати після 5 епох: 0.1989
Точність після 5 епох: 0.9268
Втрати на валідації після 5 епох: 0.3238
Точність на валідації після 5 епох: 0.8784

Модель зі шаром LSTM показує значно кращі результати порівняно з моделлю зі шаром SimpleRNN. Маючи нижчі втрати та вищу точність як на тренувальних, так і на валідаційних даних, LSTM модель демонструє кращу здатність вивчати довготривалі залежності в послідовностях та робити більш точні прогнози. LSTM має внутрішні стани та ворота, що дозволяють зберігати та забувати інформацію залежно від потреб моделі, що робить його більш потужним та гнучким для моделювання послідовностей.






